### PR TITLE
Allow arbitrary storage for `ContrastsMatrix` and other contrasts breakage

### LINKED ISSUE #175: breaking changes to contrasts
To be considered for the next breaking release (as discussed briefly in #172):

- [ ] use keyword arguments for contructors
- [ ] error on missing labels for hypothesis/contrasts coding
- [ ] more specific termnames for contrasts

### PR DESCRIPTION
**Pull request by kleinschmidt**
This makes a few breaking changes that we've planned for a while (fixes #175):

- adds an additional type parameter `M<:AbstractMatrix` to `ContrastsMatrix` which can now use any `AbstractMatrix` to represent the contrasts (with tests for `SparseMatrixCSC` storage)
- makes teh `labels=` and `levels=` kwargs for `HypothesisCoding` mandatory
- un-exports `ContrastsCoding` and removes the corresponding deprecation warning.
- removes the deprecated (and previously ignored) `base=` kwarg for `SeqDiffCoding`

I did not implement "more specific termnames" because I anticipate that being an _enormous_ bikeshed 😂

### CODE REVIEW

File: src/contrasts.jl [hunk 1]
```diff
@@ -101,19 +101,19 @@ of Memory and Language, 110_, 104038. https://doi.org/10.1016/j.jml.2019.104038
 abstract type AbstractContrasts end
 
 # Contrasts + Levels (usually from data) = ContrastsMatrix
-struct ContrastsMatrix{C <: AbstractContrasts, T, U}
-    matrix::Matrix{Float64}
+struct ContrastsMatrix{C <: AbstractContrasts, M <: AbstractMatrix, T, U}
+    matrix::M
```
Comment (nalimilan) [inline src/contrasts.jl:105]: Any particular reason to put `M` in second position? Putting it at the end would avoid breakage.

File: src/contrasts.jl [hunk 2]
```diff
@@ -553,12 +541,10 @@ mutable struct HypothesisCoding{T<:AbstractMatrix, S<:AbstractMatrix} <: Abstrac
...
-        labels == nothing && 
-            Base.depwarn(
-                "HypothesisCoding without specified contrast labels is deprecated.  " *
-                "Specify contrast labels with `HypothesisCoding(; labels=[...])` " *
-                "or HypothesisCoding(Dict(label1=>hyp1, label2=>hyp2, ...))",
-                :HypothesisCoding)
+        levels === nothing &&
+            throw(ArgumentError("kwarg `levels` must be specified for HypothesisCoding"))
+        labels === nothing &&
+            throw(ArgumentError("kwarg `labels` must be specified for HypothesisCoding"))
```
Comment (nalimilan) [inline src/contrasts.jl:547]: Maybe just dropping the default value `nothing` would be enough, as the error message is quite explicit?

**Review by nalimilan — APPROVED** [review]

File: src/contrasts.jl (hunk 1 above)
Comment (kleinschmidt) [reply src/contrasts.jl:105]: my thinking here (near as I can reconstruct it at this point) was that we're already doing a breaking release, and this position puts it in a place where you can specify it without needing to specify the other parameters.

File: test/contrasts.jl [hunk 3]
```diff
@@ -326,14 +328,14 @@
         @test_throws MethodError FullDummyCoding(levels=levs)
         @test_throws MethodError FullDummyCoding(base=base)
 
-        @test_throws ArgumentError HypothesisCoding(rand(3,4))
-        @test_throws ArgumentError HypothesisCoding(rand(3,4), levels=levs)
-        @test_throws ArgumentError HypothesisCoding(rand(3,4), labels=levs[1:3])
+        @test_throws UnderKeywordError HypothesisCoding(rand(3,4))
+        @test_throws UnderKeywordError HypothesisCoding(rand(3,4), levels=levs)
+        @test_throws UnderKeywordError HypothesisCoding(rand(3,4), labels=levs[1:3])
```
Comment (nalimilan) [inline test/contrasts.jl:333]:
```suggestion
        @test_throws UndefKeywordError HypothesisCoding(rand(3,4))
        @test_throws UndefKeywordError HypothesisCoding(rand(3,4), levels=levs)
        @test_throws UndefKeywordError HypothesisCoding(rand(3,4), labels=levs[1:3])
```

File: src/contrasts.jl (hunk 1 above)
Comment (nalimilan) [reply src/contrasts.jl:105]: OK so you suppose it will be more common to specify `M` than `T` or `U`?

Comment (kleinschmidt) [reply src/contrasts.jl:105]: yeah definitely.  those are only there to make the contrast lookup type stable/inferrable

File: test/contrasts.jl [hunk 4]
```diff
@@ -375,5 +370,31 @@
 
         @test_throws ArgumentError ContrastsMatrix(DummyCoding(levels=[1, 2, 3]), x1_levs)
     end
+
+    @testset "other matrix types" begin
+        using SparseArrays
+        contrasts = DummyCoding()
+        mat = StatsModels.contrasts_matrix(contrasts, 1, 4)
+        spmat = sparse(mat)
```
Comment (nalimilan) [inline test/contrasts.jl:378]: Unused?

Comment (kleinschmidt) [reply test/contrasts.jl]: unused

**Review by nalimilan — APPROVED** [review]

**Pull request merged**